boto/auth.py

```diff
@@ -932,11 +932,18 @@ def add_auth(self, http_request, **kwargs):
         qs = self._build_query_string(
             http_request.params
         )
-        boto.log.debug('query_string in body: %s' % qs)
-        headers['Content-Type'] = 'application/x-www-form-urlencoded'
-        # This will be  a POST so the query string should go into the body
-        # as opposed to being in the uri
-        http_request.body = qs
+        if http_request.method == 'POST':
+            boto.log.debug('query_string in body: %s' % qs)
+            headers['Content-Type'] = 'application/x-www-form-urlencoded'
+            # This is a POST so the query string should go into the body
+            # as opposed to being in the uri
+            http_request.body = qs
+        else:
+            # This is a GET so the query string should go into the URI
+            http_request.body = ''
+            http_request.path = http_request.path.split('?')[0]
```

Comment:
Hi @jgibbs-cloudian - is this split potentially throwing away query parameters that may already be in the path? Maybe it's the right thing to do, but would like to understand the logic of why it is being done.

Reply:
@tpodowd Potentially yes, though other auth handlers behave in the same way.

When forming a request, the caller provides a 'params' map and a 'path' string. My assumption (perhaps incorrect) is that callers will provide all params in the map instead of manually encoding them into the path, leaving boto to decide how to form the HTTP request.

To double check, I've looked at every caller of these 3 methods:

AWSQueryConnection.get_list
AWSQueryConnection.get_object
AWSQueryConnection.get_status

Every one leaves the path at the default ('/') apart from SQS, which sometimes sets it to an ID.  No callers manually append query params onto the path.

Each auth handler (QueryAnonHandler, QuerySignatureHelper, etc.) adds the necessary auth/signature and writes the resulting params into either the request body (for POSTs) or the path (for GETs). For the latter, the other auth handlers also remove any existing params from the path before writing it, like mine does. The others have this comment:

```
            # if this is a retried request, the qs from the previous try will
            # already be there, we need to get rid of that and rebuild it
            http_request.path = http_request.path.split('?')[0]
```

Hmm, actually, why don't I try it? If I change the path for get_caller_identity to '/?test=1' and the params to { 'Action': 'GetCallerIdentity', 'Version': '2011-06-15' }, then it appears on the wire as 'GET /?Action=GetCallerIdentity&Version=2011-06-15 HTTP/1.1' - i.e. my manually-encoded parameter is ignored. That's using one of the existing auth handlers that I haven't modified.

Given this, I'll patch my auth handler to include the same comment as the others (above). Is that sufficient?

Reply:
Sounds good to me. Makes sense. I think its good for now and if we need to change it later, we can always do so. Thanks!